metadata_layer.py

```diff
@@ -351,36 +351,188 @@ def __repr__(self):
+
+        Metadata : Edition :: CirculationData : Licensepool
+    """
+
     log = logging.getLogger(
         "Abstract metadata layer - Circulation data"
     )
 
     def __init__(
-            self, licenses_owned, 
-            licenses_available, 
-            licenses_reserved,
-            patrons_in_hold_queue,
+            self, 
+            data_source,
+            primary_identifier=None,
+            licenses_owned=None,
+            licenses_available=None,
+            licenses_reserved=None,
+            patrons_in_hold_queue=None,
+            formats=None,
             first_appearance=None,
             last_checked=None,
+            default_rights_uri=None,
+            links=None,
     ):
+        # data_source is where the lending licenses (our ability to actually access the book) are coming from.
+        self._data_source = data_source
+
+        if isinstance(self._data_source, DataSource):
+            self.data_source_obj = self._data_source
+        else:
+            self.data_source_obj = None
+
+        '''
+        if isinstance(data_source, DataSource):
+            self._data_source_obj = data_source
+            self._data_source = data_source.name
+        else:
+            self._data_source = data_source
+        '''
+
+        self.primary_identifier = primary_identifier
         self.licenses_owned = licenses_owned
         self.licenses_available = licenses_available
         self.licenses_reserved = licenses_reserved
         self.patrons_in_hold_queue = patrons_in_hold_queue
+
+        # format contains pdf/epub, drm, link
+        self.formats = formats or []
+
+        self.default_rights_uri = None
+        self.set_default_rights_uri(data_source=self.data_source, default_rights_uri=default_rights_uri)
         self.first_appearance = first_appearance
         self.last_checked = last_checked or datetime.datetime.utcnow()
 
+        self.__links = None
+        self.links = links
+        
+    @property
+    def links(self):
+        return self.__links
+
+    @links.setter
+    def links(self, arg_links):
+        """ If got passed all links, undiscriminately, filter out to only those relevant to pools. """
+        # start by deleting any old links
+        self.__links = []
+
+        if not arg_links:
+            return
+
+        for link in arg_links:
+            if link.rel in [Hyperlink.OPEN_ACCESS_DOWNLOAD, Hyperlink.DRM_ENCRYPTED_DOWNLOAD]:
+                # only accept the types of links relevant to pools
+                self.__links.append(link)
+
+                # If a link has a rights_uri, make that the overall rights_uri. 
+                # If there are multiple links with a rights_uri, they should have gotten 
+                # split into separate metadata objects in the content server or whatever code called this. 
+                if link.rights_uri:
+                    self.set_default_rights_uri(data_source=self.data_source, default_rights_uri=link.rights_uri)
```

Comment:
Since set_default_rights_uri overwrites any previous setting, you should only call it if default_rights_uri is not already set.